src/Interpreters/InterpreterDeleteQuery.cpp

```diff
@@ -83,6 +88,37 @@ BlockIO InterpreterDeleteQuery::execute()
     auto table_lock = table->lockForShare(getContext()->getCurrentQueryId(), getContext()->getSettingsRef()[Setting::lock_acquire_timeout]);
     auto metadata_snapshot = table->getInMemoryMetadataPtr();
 
+    /// DELETE FROM ... WHERE 1 should be executed as truncate
```

Comment:
The same problem with shared + exclusive lock. Here, it is easier to fix. Just do this optimization before `table->lockForShare()`.